File: go/bot.go
package chaskybot

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"regexp"
	"strings"
	"sync"
	"time"
)

// Event is one update delivered to a handler, with the shortcuts for answering
// it. Named Event rather than Context so it never reads as a context.Context.
type Event struct {
	Update  Update
	Message *Message
	ChatID  ChatID

	client   *Client
	attempts int
	onError  func(error)
}

// Reply sends to this update's chat with a managed Idempotency-Key.
func (e *Event) Reply(ctx context.Context, text string) (*Message, error) {
	return sendWithRetry(ctx, e.client, SendMessageParams{
		// G9: the chat id goes back exactly as it arrived. Never parsed, never
		// rebuilt — a real chat id carries "bot:" twice, and anything that
		// takes it apart breaks against a live server.
		ChatID: e.ChatID,
		Text:   text,
	}, e.attempts, e.onError)
}

// Typing is explicit on purpose: the runtime never sends a chat action by
// itself. Whether the indicator helps depends on how long the bot takes, which
// is the author's call, not the SDK's.
func (e *Event) Typing(ctx context.Context) error {
	return e.client.SendChatAction(ctx, e.ChatID, ActionTyping)
}

type Handler func(ctx context.Context, e *Event) error

type BotOptions struct {
	Options
	Transport Transport
	// SendAttempts bounds the retries of one logical send. All of them reuse
	// the same Idempotency-Key.
	SendAttempts int
}

type Bot struct {
	client    *Client
	transport Transport
	attempts  int

	mu       sync.Mutex
	running  bool
	commands map[string]Handler
	handlers []Handler
	onErrors []func(error)
}

func New(opts BotOptions) (*Bot, error) {
	client, err := NewClient(opts.Options)
	if err != nil {
		return nil, err
	}
	transport := opts.Transport
	if transport == nil {
		transport = Polling(PollingOptions{})
	}
	attempts := opts.SendAttempts
	if attempts <= 0 {
		attempts = 4
	}
	return &Bot{client: client, transport: transport, attempts: attempts, commands: map[string]Handler{}}, nil
}

func (b *Bot) Client() *Client { return b.client }

func (b *Bot) Command(name string, h Handler) *Bot {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.commands[strings.TrimPrefix(name, "/")] = h
	return b
}

// Handle registers a handler for any text message that matched no command.
func (b *Bot) Handle(h Handler) *Bot {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.handlers = append(b.handlers, h)
	return b
}

func (b *Bot) OnError(fn func(error)) *Bot {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.onErrors = append(b.onErrors, fn)
	return b
}

// ErrAlreadyRunning guards G1: exactly one in-flight poll per instance. Two
// overlapping polls on one bot used to split the updates silently; the server
// now answers 409, but the SDK should never be the one causing it.
var ErrAlreadyRunning = errors.New("chaskybot: this bot is already running")

// Run blocks until ctx is cancelled, which returns nil, or until a terminal
// failure such as 401, 403 or 409 CONFLICT_POLLING, which is returned.
func (b *Bot) Run(ctx context.Context) error {
	b.mu.Lock()
	if b.running {
		b.mu.Unlock()
		return ErrAlreadyRunning
	}
	b.running = true
	b.mu.Unlock()
	defer func() {
		b.mu.Lock()
		b.running = false
		b.mu.Unlock()
	}()

	return b.transport.Run(ctx, TransportDeps{
		Client:   b.client,
		OnUpdate: b.dispatch,
		OnError:  b.emitError,
	})
}

func (b *Bot) emitError(err error) {
	b.mu.Lock()
	listeners := append([]func(error){}, b.onErrors...)
	b.mu.Unlock()
	for _, fn := range listeners {
		fn(err)
	}
}

func (b *Bot) dispatch(ctx context.Context, update Update) error {
	if update.Message == nil {
		return nil
	}
	event := &Event{
		Update:   update,
		Message:  update.Message,
		ChatID:   update.Message.Chat.ID,
		client:   b.client,
		attempts: b.attempts,
		onError:  b.emitError,
	}

	b.mu.Lock()
	handler, isCommand := b.commands[parseCommand(update.Message.Text)]
	fallbacks := append([]Handler{}, b.handlers...)
	b.mu.Unlock()

	if isCommand {
		return handler(ctx, event)
	}
	for _, fn := range fallbacks {
		if err := fn(ctx, event); err != nil {
			return err
		}
	}
	return nil
}

// sendWithRetry implements G4: one key per logical message, the SAME key on
// every retry of THAT message.
//
// The key is generated once, outside the loop. A fresh key per attempt turns a
// retry into a second message and the user sees the bot answer twice; reusing
// one key across different messages makes the server discard the second
// SILENTLY, which is worse because it looks like it worked.
func sendWithRetry(ctx context.Context, c *Client, p SendMessageParams, attempts int, onError func(error)) (*Message, error) {
	key, err := newIdempotencyKey()
	if err != nil {
		return nil, err
	}
	p.IdempotencyKey = key

	var last error
	for attempt := 1; attempt <= attempts; attempt++ {
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		message, err := c.SendMessage(ctx, p)
		if err == nil {
			return message, nil
		}
		last = err
		if classify(err, sourceCall) != classTransient {
			return nil, err
		}
		if attempt == attempts {
			break
		}
		if onError != nil {
			onError(err)
		}
		if !sleepCtx(ctx, backoff(attempt, 250*time.Millisecond, 4*time.Second)) {
			return nil, ctx.Err()
		}
	}
	return nil, last
}

func newIdempotencyKey() (string, error) {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		// Without randomness a unique key cannot be produced, and reusing one
		// makes the server drop the message without saying so. Not sending is
		// the safer failure: it is visible.
		return "", &UsageError{Message: "no secure randomness available to build an Idempotency-Key; refusing to send"}
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	s := hex.EncodeToString(b[:])
	return s[:8] + "-" + s[8:12] + "-" + s[12:16] + "-" + s[16:20] + "-" + s[20:], nil
}

var commandPattern = regexp.MustCompile(`^/([A-Za-z0-9_]+)`)

func parseCommand(text string) string {
	match := commandPattern.FindStringSubmatch(strings.TrimSpace(text))
	if match == nil {
		return ""
	}
	return match[1]
}

File: js/src/bot.ts
import { ChaskyBotClient, type ChaskyBotClientOptions } from "./client.js";
import { ChaskyUsageError, classify } from "./errors.js";
import { polling, sleep, type Transport } from "./polling.js";
import type { BotMessage, ChatID, MessageID, Update } from "./types.js";

export interface Context {
  readonly update: Update;
  readonly message: BotMessage;
  readonly chatId: ChatID;
  readonly client: ChaskyBotClient;
  /** sendMessage to this update's chat, with a managed Idempotency-Key. */
  reply(text: string, options?: { replyToMessageId?: MessageID }): Promise<BotMessage>;
  /** Explicit on purpose — the runtime never sends a chat action by itself. */
  typing(): Promise<void>;
}

export type Handler = (ctx: Context) => void | Promise<void>;

export interface CreateBotOptions extends ChaskyBotClientOptions {
  transport?: Transport | undefined;
  /** Attempts for one logical send before giving up. All reuse the same key. */
  sendAttempts?: number | undefined;
}

export interface Bot {
  command(name: string, handler: Handler): Bot;
  on(event: "text", handler: Handler): Bot;
  onError(listener: (error: unknown) => void): Bot;
  onFatal(listener: (error: unknown) => void): Bot;
  start(): Promise<void>;
  stop(): Promise<void>;
  readonly client: ChaskyBotClient;
}

export function createBot(options: CreateBotOptions): Bot {
  const client = new ChaskyBotClient(options);

  const transport = options.transport ?? polling();
  const sendAttempts = options.sendAttempts ?? 4;
  const commands = new Map<string, Handler>();
  const textHandlers: Handler[] = [];
  const errorListeners: ((error: unknown) => void)[] = [];
  const fatalListeners: ((error: unknown) => void)[] = [];

  let controller: AbortController | null = null;
  let loop: Promise<void> | null = null;

  const emitError = (error: unknown): void => {
    for (const listener of errorListeners) listener(error);
  };
  const emitFatal = (error: unknown): void => {
    for (const listener of fatalListeners) listener(error);
  };

  async function dispatch(update: Update): Promise<void> {
    const message = update.message;
    if (!message) return;
    const ctx = makeContext(update, message);
    const command = parseCommand(message.text);
    const handler = command ? commands.get(command) : undefined;
    if (handler) {
      await handler(ctx);
      return;
    }
    for (const textHandler of textHandlers) await textHandler(ctx);
  }

  function makeContext(update: Update, message: BotMessage): Context {
    return {
      update,
      message,
      chatId: message.chat.id,
      client,
      reply: (text, replyOptions) =>
        sendWithRetry(
          client,
          // G9: the chat id goes back exactly as it arrived. Never parsed,
          // never coerced — a coerced id addresses the wrong chat silently.
          { chatId: message.chat.id, text, ...(replyOptions?.replyToMessageId ? { replyToMessageId: replyOptions.replyToMessageId } : {}) },
          { attempts: sendAttempts, signal: controller?.signal, onError: emitError },
        ),
      typing: async () => {
        await client.sendChatAction({ chatId: message.chat.id, action: "typing" }, controller?.signal);
      },
    };
  }

  const bot: Bot = {
    client,
    command(name, handler) {
      commands.set(name.replace(/^\//, ""), handler);
      return bot;
    },
    on(_event, handler) {
      textHandlers.push(handler);
      return bot;
    },
    onError(listener) {
      errorListeners.push(listener);
      return bot;
    },
    onFatal(listener) {
      fatalListeners.push(listener);
      return bot;
    },
    async start() {
      // G1: one in-flight poll per instance. Two overlapping polls on one bot
      // used to split the updates silently; the server now answers 409, but the
      // SDK should never be the one causing it.
      if (controller) throw new ChaskyUsageError("this bot is already started; call stop() before starting again");
      controller = new AbortController();
      const signal = controller.signal;
      loop = transport
        .run({ client, onUpdate: dispatch, onError: emitError, onFatal: emitFatal, signal })
        .catch((error: unknown) => {
          // A transport that throws out of its own loop is a bug in the SDK, not
          // in the author's bot. It must still reach them rather than vanish
          // into an unhandled rejection.
          emitFatal(error);
        });
    },
    async stop() {
      // G8: abort the in-flight long poll instead of waiting out the server's
      // deadline, which can be 30 seconds away.
      controller?.abort();
      controller = null;
      await loop?.catch(() => {});
      loop = null;
    },
  };

  return bot;
}

/**
 * G4: one key per logical message, the SAME key on every retry of THAT message.
 *
 * The key is generated once, outside the retry loop. Generating a fresh key per
 * attempt would turn a retry into a second message — the user sees the bot
 * answer twice — and reusing one key across different messages makes the server
 * discard the second SILENTLY, which is worse because it looks like it worked.
 */
async function sendWithRetry(
  client: ChaskyBotClient,
  params: { chatId: ChatID; text: string; replyToMessageId?: MessageID },
  options: { attempts: number; signal?: AbortSignal | undefined; onError: (error: unknown) => void },
): Promise<BotMessage> {
  const idempotencyKey = newIdempotencyKey();
  let lastError: unknown;
  for (let attempt = 1; attempt <= options.attempts; attempt++) {
    if (options.signal?.aborted) break;
    try {
      return await client.sendMessage(params, { idempotencyKey, ...(options.signal ? { signal: options.signal } : {}) });
    } catch (error) {
      lastError = error;
      if (classify(error, "call") !== "transient") throw error;
      if (attempt === options.attempts) break;
      options.onError(error);
      await sleep(Math.round(250 * 2 ** (attempt - 1) * (0.5 + Math.random() * 0.5)), options.signal);
    }
  }
  throw lastError;
}

function newIdempotencyKey(): string {
  const uuid = globalThis.crypto?.randomUUID?.();
  if (!uuid) {
    // Without randomness a unique key cannot be produced, and reusing one makes
    // the server drop the message without saying so. Not sending is the safer
    // failure: it is visible.
    throw new ChaskyUsageError("no secure randomness available to build an Idempotency-Key; refusing to send");
  }
  return uuid;
}

function parseCommand(text: string): string | undefined {
  const match = /^\/([A-Za-z0-9_]+)/.exec(text.trim());
  return match?.[1];
}
